File: ctim.cpp
#include <cassert>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <optional>
#include <regex>
#include <sstream>
#include <string>
#include <tuple>
#include <type_traits>

std::optional<std::string>
encodeCTIM(uint32_t ledger_seq, uint16_t txn_index, uint16_t network_id) noexcept
{
  if (ledger_seq > 0xFFFFFFF)
    return {};

  if (txn_index > 0xFFFF)
    return {};

  if (network_id > 0xFFFF)
    return {};

  uint64_t ctimValue =
      ((0xC0000000ULL + static_cast<uint64_t>(ledger_seq)) << 32) +
      (static_cast<uint64_t>(txn_index) << 16) + network_id;

  std::stringstream buffer;
  buffer << std::hex << std::uppercase << std::setfill('0') << std::setw(16)
         << ctimValue;
  return {buffer.str()};
}

template <typename T>
std::optional<std::tuple<uint32_t, uint16_t, uint16_t>>
decodeCTIM(const T ctim) noexcept
{
  uint64_t ctimValue {0};
  if constexpr (std::is_same_v<T, std::string> || std::is_same_v<T, char *> ||
                std::is_same_v<T, const char *> ||
                std::is_same_v<T, std::string_view>)
  {
    const std::string ctimString(ctim);

    if (ctimString.length() != 16)
      return {};

    if (!std::regex_match(ctimString, std::regex("^[0-9A-F]+$")))
      return {};

    ctimValue = std::stoull(ctimString, nullptr, 16);
  } else if constexpr (std::is_integral_v<T>)
    ctimValue = ctim;
  else
    return {};

  if (ctimValue > 0xFFFFFFFFFFFFFFFFULL ||
      (ctimValue & 0xF000000000000000ULL) != 0xC000000000000000ULL)
    return {};

  uint32_t ledger_seq = (ctimValue >> 32) & 0xFFFFFFFUL;
  uint16_t txn_index = (ctimValue >> 16) & 0xFFFFU;
  uint16_t network_id = ctimValue & 0xFFFFU;
  return {{ledger_seq, txn_index, network_id}};
}

// NOTE TO DEVELOPER:
// you only need the two functions above, below are test cases for nodejs, if
// you want them.

int main() {
  std::cout << "Running test cases..." << std::endl;
  // Test case 1: Valid input values
  assert(encodeCTIM(0xFFFFFFFUL, 0xFFFFU, 0xFFFFU) ==
         std::optional<std::string>("CFFFFFFFFFFFFFFF"));
  assert(encodeCTIM(0, 0, 0) == std::optional<std::string>("C000000000000000"));
  assert(encodeCTIM(1U, 2U, 3U) ==
         std::optional<std::string>("C000000100020003"));
  assert(encodeCTIM(13249191UL, 12911U, 49221U) ==
         std::optional<std::string>("C0CA2AA7326FC045"));

  // Test case 2: ledger_seq greater than 0xFFFFFFF
  assert(!encodeCTIM(0x10000000UL, 0xFFFFU, 0xFFFFU));

  // Test case 3: txn_index greater than 0xFFFF
  // this test case is impossible in c++ due to the type, left in for
  // completeness assert(!encodeCTIM(0xFFFFFFF, 0x10000, 0xFFFF));

  // Test case 4: network_id greater than 0xFFFF
  // this test case is impossible in c++ due to the type, left in for
  // completeness assert(!encodeCTIM(0xFFFFFFFUL, 0xFFFFU, 0x10000U));

  // Test case 5: Valid input values
  assert((decodeCTIM("CFFFFFFFFFFFFFFF") ==
          std::optional<std::tuple<int32_t, uint16_t, uint16_t>>(
              std::make_tuple(0xFFFFFFFULL, 0xFFFFU, 0xFFFFU))));
  assert((decodeCTIM("C000000000000000") ==
          std::optional<std::tuple<int32_t, uint16_t, uint16_t>>(
              std::make_tuple(0, 0, 0))));
  assert((decodeCTIM("C000000100020003") ==
          std::optional<std::tuple<int32_t, uint16_t, uint16_t>>(
              std::make_tuple(1U, 2U, 3U))));
  assert((decodeCTIM("C0CA2AA7326FC045") ==
          std::optional<std::tuple<int32_t, uint16_t, uint16_t>>(
              std::make_tuple(13249191UL, 12911U, 49221U))));

  // Test case 6: ctim not a string or big int
  assert(!decodeCTIM(0xCFF));

  // Test case 7: ctim not a hexadecimal string
  assert(!decodeCTIM("C003FFFFFFFFFFFG"));

  // Test case 8: ctim not exactly 16 nibbles
  assert(!decodeCTIM("C003FFFFFFFFFFF"));

  // Test case 9: ctim too large to be a valid CTIM value
  assert(!decodeCTIM("CFFFFFFFFFFFFFFFF"));

  // Test case 10: ctim doesn't start with a C nibble
  assert(!decodeCTIM("FFFFFFFFFFFFFFFF"));

  // Test case 11: Valid input values
  assert((decodeCTIM(0xCFFFFFFFFFFFFFFFULL) ==
          std::optional<std::tuple<int32_t, uint16_t, uint16_t>>(
              std::make_tuple(0xFFFFFFFUL, 0xFFFFU, 0xFFFFU))));
  assert((decodeCTIM(0xC000000000000000ULL) ==
          std::optional<std::tuple<int32_t, uint16_t, uint16_t>>(
              std::make_tuple(0, 0, 0))));
  assert((decodeCTIM(0xC000000100020003ULL) ==
          std::optional<std::tuple<int32_t, uint16_t, uint16_t>>(
              std::make_tuple(1U, 2U, 3U))));
  assert((decodeCTIM(0xC0CA2AA7326FC045ULL) ==
          std::optional<std::tuple<int32_t, uint16_t, uint16_t>>(
              std::make_tuple(13249191UL, 12911U, 49221U))));

  // Test case 12: ctim not exactly 16 nibbles
  assert(!decodeCTIM(0xC003FFFFFFFFFFF));

  // Test case 13: ctim too large to be a valid CTIM value
  // this test case is not possible in c++ because it would overflow the type,
  // left in for completeness assert(!decodeCTIM(0xCFFFFFFFFFFFFFFFFULL));

  // Test case 14: ctim doesn't start with a C nibble
  assert(!decodeCTIM(0xFFFFFFFFFFFFFFFFULL));

  std::cout << "Done!" << std::endl;
  return 0;
}

File: ctim.php
<?php

function encodeCTIM($lgrIndex, $txnIndex, $networkId)
{
  if (!is_numeric($lgrIndex))
    throw new Exception("lgrIndex must be a number.");
  if ($lgrIndex > 0xFFFFFFF || $lgrIndex < 0)
    throw new Exception("lgrIndex must not be greater than 268435455 or less than 0.");

  if (!is_numeric($txnIndex))
    throw new Exception("txnIndex must be a number.");
  if ($txnIndex > 0xFFFF || $txnIndex < 0)
    throw new Exception("txnIndex must not be greater than 65535 or less than 0.");

  if (!is_numeric($networkId))
    throw new Exception("networkId must be a number.");
  if ($networkId > 0xFFFF || $networkId < 0)
    throw new Exception("networkId must not be greater than 65535 or less than 0.");

  $ledger_part = dechex($lgrIndex);
  $txn_part = dechex($txnIndex);
  $network_part = dechex($networkId);

  if (strlen($ledger_part) < 7)
      $ledger_part = str_repeat("0", 7 - strlen($ledger_part)) . $ledger_part;
  if (strlen($txn_part) < 4)
      $txn_part = str_repeat("0", 4 - strlen($txn_part)) . $txn_part;
  if (strlen($network_part) < 4)
      $network_part = str_repeat("0", 4 - strlen($network_part)) . $network_part;

  return strtoupper("C" . $ledger_part . $txn_part . $network_part);
}

function decodeCTIM($ctim)
{
  if (is_string($ctim))
  {
    if (!ctype_xdigit($ctim))
      throw new Exception("ctim must be a hexadecimal string");
    if (strlen($ctim) !== 16)
      throw new Exception("ctim must be exactly 16 nibbles and start with a C");
  } else
    throw new Exception("ctim must be a hexadecimal string");

  if (substr($ctim, 0, 1) !== 'C')
    throw new Exception("ctim must be exactly 16 nibbles and start with a C");

  $lgrIndex = substr($ctim, 1, 7);
  $txnIndex = substr($ctim, 8, 4);
  $networkId = substr($ctim, 12, 4);
  return array(
    "lgrIndex" => hexdec($lgrIndex),
    "txnIndex" => hexdec($txnIndex),
    "networkId" => hexdec($networkId)
  );
}

// NOTE TO DEVELOPER:
// you only need the two functions above, below are test cases, if you want them.

print("Running tests...\n");

function assert_test($x)
{
    if (!$x)
        echo "test failed!\n";
    else
        echo "test passed\n";
}

// Test case 1: Valid input values
assert_test(encodeCTIM(0xFFFFFFF, 0xFFFF, 0xFFFF) == "CFFFFFFFFFFFFFFF");
assert_test(encodeCTIM(0, 0, 0) == "C000000000000000");
assert_test(encodeCTIM(1, 2, 3) == "C000000100020003");
assert_test(encodeCTIM(13249191, 12911, 49221) == "C0CA2AA7326FC045");

// Test case 2: lgrIndex greater than 0xFFFFFFF
try {
  encodeCTIM(0x10000000, 0xFFFF, 0xFFFF);
  assert_test(false);
} catch (Exception $e) {
  assert_test(strcmp($e->getMessage(), "lgrIndex must not be greater than 268435455 or less than 0.") == 0);
}
try {
  encodeCTIM(-1, 0xFFFF, 0xFFFF);
  assert_test(false);
} catch (Exception $e) {
  assert_test(strcmp($e->getMessage(), "lgrIndex must not be greater than 268435455 or less than 0.") == 0);
}

// Test case 3: txnIndex greater than 0xFFFF
try {
  encodeCTIM(0xFFFFFFF, 0x10000, 0xFFFF);
  assert_test(false);
} catch (Exception $e) {
  assert_test(strcmp($e->getMessage(), "txnIndex must not be greater than 65535 or less than 0.") == 0);
}
try {
  encodeCTIM(0xFFFFFFF, -1, 0xFFFF);
  assert_test(false);
} catch (Exception $e) {
  assert_test(strcmp($e->getMessage(), "txnIndex must not be greater than 65535 or less than 0.") == 0);
}

// Test case 4: networkId greater than 0xFFFF
try {
  encodeCTIM(0xFFFFFFF, 0xFFFF, 0x10000);
  assert_test(false);
} catch (Exception $e) {
  assert_test(strcmp($e->getMessage(), "networkId must not be greater than 65535 or less than 0.") == 0);
}
try {
  encodeCTIM(0xFFFFFFF, 0xFFFF, -1);
  assert_test(false);
} catch (Exception $e) {
  assert_test(strcmp($e->getMessage(), "networkId must not be greater than 65535 or less than 0.") == 0);
}

// Test case 5: Valid input values
assert_test(decodeCTIM("CFFFFFFFFFFFFFFF") == array("lgrIndex" => 0xFFFFFFF, "txnIndex" => 0xFFFF, "networkId" => 0xFFFF));
assert_test(decodeCTIM("C000000000000000") == array("lgrIndex" => 0, "txnIndex" => 0, "networkId" => 0));
assert_test(decodeCTIM("C000000100020003") == array("lgrIndex" =>1, "txnIndex" => 2, "networkId" => 3));
assert_test(decodeCTIM("C0CA2AA7326FC045") == array("lgrIndex" =>13249191, "txnIndex" => 12911, "networkId" => 49221));


print("Done!\n");

?>
